url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/plugins/xml/XMLSchedulingDataProcessor.java

```diff
@@ -0,0 +1,592 @@
+      }
+    }
+
+    if (!fileFound) {
+      if (failOnFileNotFound) {
+        throw new SchedulerException("File named '" + fileName + "' does not exist.");
+      } else {
+        logger.warn("File named '" + fileName + "' does not exist. This is OK if you don't want to use an XML job config file.");
+      }
+    } else {
+      processFile(fileName);
+    }
+  }
+
+  /**
+   * Process the xmlfile named <code>fileName</code> with the given system ID.
+   *
+   * @param fileName meta data file name.
+   * @param systemId system ID.
+   */
+  private void processFile(String fileName) throws ValidationException, ParserConfigurationException, SAXException, IOException, SchedulerException,
+      ClassNotFoundException, ParseException, XPathException {
+
+    prepForProcessing();
+
+    logger.info("Parsing XML file: " + fileName);
+    InputSource is = new InputSource(getInputStream(fileName));
+    // is.setSystemId(systemId);
+
+    process(is);
+
+    maybeThrowValidationException();
+  }
+
+  /*
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Interface.
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
+
+  private void prepForProcessing() {
+
+    clearValidationExceptions();
+
+    loadedJobs.clear();
+    loadedTriggers.clear();
+  }
+
+  private void process(InputSource is) throws SAXException, IOException, ParseException, XPathException, ClassNotFoundException {
+
+    // load the document
+    Document document = docBuilder.parse(is);
+
+    //
+    // Extract Job definitions...
+    //
+
+    NodeList jobNodes = (NodeList) xpath.evaluate("/job-scheduling-data/schedule/job", document, XPathConstants.NODESET);
+
+    logger.debug("Found " + jobNodes.getLength() + " job definitions.");
+
+    for (int i = 0; i < jobNodes.getLength(); i++) {
+
+      Node jobDetailNode = jobNodes.item(i);
+
+      String jobName = getTrimmedToNullString(xpath, "name", jobDetailNode);
+      String jobDescription = getTrimmedToNullString(xpath, "description", jobDetailNode);
+      String jobClassName = getTrimmedToNullString(xpath, "job-class", jobDetailNode);
+      boolean isConcurrencyAllowed = getBoolean(xpath, "concurrency-allowed", jobDetailNode);
+      Class<? extends Job> jobClass = classLoadHelper.loadClass(jobClassName);
+
+      JobDetail jobDetail = newJobBuilder(jobClass).withIdentity(jobName).isConcurrencyAllowed(isConcurrencyAllowed).withDescription(jobDescription)
+          .build();
+
+      NodeList jobDataEntries = (NodeList) xpath.evaluate("job-data-map/entry", jobDetailNode, XPathConstants.NODESET);
+
+      for (int k = 0; k < jobDataEntries.getLength(); k++) {
+        Node entryNode = jobDataEntries.item(k);
+        String key = getTrimmedToNullString(xpath, "key", entryNode);
+        String value = getTrimmedToNullString(xpath, "value", entryNode);
+        jobDetail.getJobDataMap().put(key, value);
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 301:306, 411:416.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.